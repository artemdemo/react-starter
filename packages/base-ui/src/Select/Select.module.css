.Select.form-control {
  width: auto;
  max-width: 100%;
  display: inline-block;
}
