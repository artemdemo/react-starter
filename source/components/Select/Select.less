.select {
    width: auto;
    max-width: 100%;
    display: inline-block;
}
