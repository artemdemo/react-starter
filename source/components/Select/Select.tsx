import React from 'react';
import classnames from 'classnames';
import _get from 'lodash/get';
import styled from 'styled-components';

const SelectContainer = styled.select`
  width: auto;
  max-width: 100%;
  display: inline-block;
`;

const PLACEHOLDER_VALUE = 'placeholder-12#$';

export enum ESelectSize {
  large,
  small,
}

export type TSelectListItem = {
  value: string;
  name: string;
  disabled?: boolean;
};

type TProps = {
  className: string;
  id: string;
  placeholder: string;
  disabled: boolean;
  list: TSelectListItem[];
  onChange?: (TSelectListItem) => void;
  value: TSelectListItem;
  size?: ESelectSize;
};

export const Select = (props: TProps) => {
  const {className, id, placeholder, disabled, list, onChange, value, size} = props;
  const onChangeValue = (e) => {
    onChange && onChange(list.find(item => item.value === e.target.value));
  };
  const selectedValue = value && list.find((item) => {
    return item === value;
  });
  const renderPlaceholder = () => {
    if (placeholder) {
      return (
        <option
          value={PLACEHOLDER_VALUE}
          disabled
        >
          {placeholder}
        </option>
      );
    }
    return null;
  };

  return (
    <SelectContainer
      className={classnames(className, {
        'form-control': true,
        'form-control-lg': size === ESelectSize.large,
        'form-control-sm': size === ESelectSize.small,
      })}
      defaultValue={placeholder && !value ? PLACEHOLDER_VALUE : undefined}
      value={_get(selectedValue, 'value', undefined)}
      onChange={onChange && onChangeValue}
      disabled={disabled}
      id={id}
    >
      {renderPlaceholder()}
      {list.map(item => (
        <option
          value={item.value}
          disabled={!item.value || item.disabled}
          key={`${item.value} - ${item.name}`}
        >
          {item.name}
        </option>
      ))}
    </SelectContainer>
  );
};

Select.defaultProps = {
  className: '',
  id: undefined,
  placeholder: undefined,
  disabled: false,
  list: [],
  value: undefined,
};
