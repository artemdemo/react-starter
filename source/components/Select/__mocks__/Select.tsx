import React from 'react';

const Select = props => (
    <div data-mock='Select'>
        {JSON.stringify(props, null, 2)}
    </div>
);

export default Select;
