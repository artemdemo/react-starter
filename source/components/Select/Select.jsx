import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import _get from 'lodash/get';

import './Select.less';

const PLACEHOLDER_VALUE = 'placeholder-12#$';

const Select = (props) => {
    const { className, id, placeholder, disabled, list, onChange, value, large, small } = props;
    const onChangeValue = (e) => {
        onChange(list.find(item => item.value === e.target.value));
    };
    const selectedValue = value && list.find((item) => {
        return item === value;
    });
    const renderPlaceholder = () => {
        if (placeholder) {
            return (
                <option
                    value={PLACEHOLDER_VALUE}
                    disabled
                >
                    {placeholder}
                </option>
            );
        }
        return null;
    };

    return (
        <select
            className={classnames(className, {
                select: true,
                'form-control': true,
                'form-control-lg': large,
                'form-control-sm': small,
            })}
            defaultValue={placeholder && !value ? PLACEHOLDER_VALUE : undefined}
            value={_get(selectedValue, 'value', undefined)}
            onChange={onChange && onChangeValue}
            disabled={disabled}
            id={id}
        >
            {renderPlaceholder()}
            {list.map(item => (
                <option
                    value={item.value}
                    disabled={!item.value || item.disabled}
                    key={`${item.value} - ${item.name}`}
                >
                    {item.name}
                </option>
            ))}
        </select>
    );
};

const valueProp = PropTypes.shape({
    value: PropTypes.string,
    name: PropTypes.string,
    disabled: PropTypes.bool,
});

Select.propTypes = {
    className: PropTypes.string,
    id: PropTypes.string,
    placeholder: PropTypes.string,
    disabled: PropTypes.bool,
    list: PropTypes.arrayOf(valueProp),
    value: valueProp,
    onChange: PropTypes.func,
    large: PropTypes.bool,
    small: PropTypes.bool,
};

Select.defaultProps = {
    className: '',
    id: undefined,
    placeholder: undefined,
    disabled: false,
    list: [],
    value: undefined,
    onChange: undefined,
    large: false,
    small: false,
};

export default Select;
