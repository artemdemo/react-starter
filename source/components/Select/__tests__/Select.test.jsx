import React from 'react';
import renderer from 'react-test-renderer';
import Select from '../Select';

describe('<Select>', () => {
    it('Simple render', () => {
        const tree = renderer.create(
            <Select />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('large', () => {
        const tree = renderer.create(
            <Select large />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('small', () => {
        const tree = renderer.create(
            <Select small />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('with className', () => {
        const tree = renderer.create(
            <Select className='some-class' />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('with id', () => {
        const tree = renderer.create(
            <Select id='some-id' />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('with items', () => {
        const tree = renderer.create(
            <Select
                list={[
                    {value: 'first', name: 'First'},
                    {value: 'last', name: 'Last'},
                ]}
            />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('with items and all disabled', () => {
        const tree = renderer.create(
            <Select
                list={[
                    {value: 'first', name: 'First'},
                    {value: 'last', name: 'Last'},
                ]}
                disabled
            />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('with items and first explicetly disabled', () => {
        const tree = renderer.create(
            <Select
                list={[
                    {value: 'first', name: 'First', disabled: true},
                    {value: 'last', name: 'Last'},
                ]}
                disabled
            />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('with items - first without a value', () => {
        const tree = renderer.create(
            <Select
                list={[
                    {name: 'First'},
                    {value: 'last', name: 'Last'},
                ]}
            />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
