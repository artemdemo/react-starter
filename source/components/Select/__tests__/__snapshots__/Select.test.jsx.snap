// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select> Simple render 1`] = `
<select
  className="select form-control"
  disabled={false}
/>
`;

exports[`<Select> large 1`] = `
<select
  className="select form-control form-control-lg"
  disabled={false}
/>
`;

exports[`<Select> small 1`] = `
<select
  className="select form-control form-control-sm"
  disabled={false}
/>
`;

exports[`<Select> with className 1`] = `
<select
  className="some-class select form-control"
  disabled={false}
/>
`;

exports[`<Select> with id 1`] = `
<select
  className="select form-control"
  disabled={false}
  id="some-id"
/>
`;

exports[`<Select> with items - first without a value 1`] = `
<select
  className="select form-control"
  disabled={false}
>
  <option
    disabled={true}
  >
    First
  </option>
  <option
    value="last"
  >
    Last
  </option>
</select>
`;

exports[`<Select> with items 1`] = `
<select
  className="select form-control"
  disabled={false}
>
  <option
    value="first"
  >
    First
  </option>
  <option
    value="last"
  >
    Last
  </option>
</select>
`;

exports[`<Select> with items and all disabled 1`] = `
<select
  className="select form-control"
  disabled={true}
>
  <option
    value="first"
  >
    First
  </option>
  <option
    value="last"
  >
    Last
  </option>
</select>
`;

exports[`<Select> with items and first explicetly disabled 1`] = `
<select
  className="select form-control"
  disabled={true}
>
  <option
    disabled={true}
    value="first"
  >
    First
  </option>
  <option
    value="last"
  >
    Last
  </option>
</select>
`;
